### PR TITLE
GrpcJsonBridge

### PR DESCRIPTION
**Pull request by jendakol**
Currently a draft of the trait. Once we will agree on it, I will implement macros to generate the code automatically.

### CODE REVIEW

**Comment by jendakol** [comment]
@augi @petr-k Ping, guys :-)

File: core/src/main/scala/com/avast/grpc/jsonbridge/GrpcJsonBridge.scala [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.avast.grpc.jsonbridge
+
+import io.grpc.BindableService
+
+import scala.concurrent.Future
+
+trait GrpcJsonBridge[Service <: BindableService] {
+  def invokeGrpcMethod(name: String, json: String, headers: Seq[GrpcHeader] = Seq.empty): Option[Future[String]]
```
Comment (augi) [inline core/src/main/scala/com/avast/grpc/jsonbridge/GrpcJsonBridge.scala]: What is the meaning of `Option[Future[String]]`? Please add a comment.

File: core/src/test/java/com/avast/grpc/jsonbridge/test/TestApi.java [hunk 2]
```diff
@@ -0,0 +1,1762 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (augi) [inline core/src/test/java/com/avast/grpc/jsonbridge/test/TestApi.java]: Please don't commit generated code.

File: core/src/test/java/com/avast/grpc/jsonbridge/test/TestApiOrBuilder.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (augi) [inline core/src/test/java/com/avast/grpc/jsonbridge/test/TestApiOrBuilder.java]: Please don't commit generated code.

File: core/src/main/scala/com/avast/grpc/jsonbridge/GrpcJsonBridgeBase.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package com.avast.grpc.jsonbridge
+
+import com.google.protobuf.Message
+import com.google.protobuf.util.JsonFormat
+import io.grpc.Metadata
+import io.grpc.stub.MetadataUtils
+
+import scala.concurrent.Future
+
+trait GrpcJsonBridgeBase[Stub <: io.grpc.stub.AbstractStub[Stub]] {
```
Comment (augi) [inline core/src/main/scala/com/avast/grpc/jsonbridge/GrpcJsonBridgeBase.scala]: I tend to use `abstract class` for `Base` classes.

File: core/src/main/scala/com/avast/grpc/jsonbridge/GrpcJsonBridgeBase.scala [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+  protected def newFutureStub: Stub
+
+  // https://groups.google.com/forum/#!topic/grpc-io/1-KMubq1tuc
+  protected def withNewFutureStub[A](headers: Seq[GrpcHeader])(f: Stub => Future[A]): Future[A] = {
```
Comment (augi) [inline core/src/main/scala/com/avast/grpc/jsonbridge/GrpcJsonBridgeBase.scala]: Maybe the name could reflect that it adds headers, like `withHeaders`.

File: core/src/main/scala/com/avast/grpc/jsonbridge/GrpcJsonBridge.scala (hunk 1 above)
Comment (jendakol) [reply core/src/main/scala/com/avast/grpc/jsonbridge/GrpcJsonBridge.scala]: ok

File: core/src/main/scala/com/avast/grpc/jsonbridge/GrpcJsonBridgeBase.scala (hunk 4 above)
Comment (jendakol) [reply core/src/main/scala/com/avast/grpc/jsonbridge/GrpcJsonBridgeBase.scala]: Don't see a reason for that.. :-)

**Comment by jendakol** [comment]
@augi Configured Java test-sources compilation.

**Comment by jendakol** [comment]
I'm going to implement macros generating the code I've introduced.

**Comment by jendakol** [comment]
@augi @petr-k Macros have been implemented, incl. Cactus support. See enclosed test for usage.  
The project is missing readme which I will add after all implementation is done.

Question is whether we don't want to have support for some _service registry_ or we state that having multiple gRPC services in one application is very major case and thus it's not worth it.

What's following is implementation for some HTTP servers - we have to create a list of them.

**Pull request merged**